File: haven-desktop/engine/indicators/indicators.go
// Package indicators provides technical analysis indicator functions.
// All functions operate on float64 slices and return null (NaN) for
// warm-up periods where there isn't enough data.
//
// These are the Go-native implementations used by the backtest engine.
// The same functions are exposed to user JavaScript strategies via goja bindings.
package indicators

import "math"

// SMA returns the Simple Moving Average over the last `period` values.
// Returns NaN for indices < period-1.
func SMA(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	sum := 0.0
	for i, v := range data {
		sum += v
		if i >= period {
			sum -= data[i-period]
		}
		if i >= period-1 {
			out[i] = sum / float64(period)
		} else {
			out[i] = math.NaN()
		}
	}
	return out
}

// EMA returns the Exponential Moving Average.
// Uses SMA for the initial seed value, then 2/(period+1) smoothing.
func EMA(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	k := 2.0 / float64(period+1)

	// Seed with SMA
	seedSum := 0.0
	for i := 0; i < period && i < len(data); i++ {
		seedSum += data[i]
		out[i] = math.NaN()
	}
	if len(data) < period {
		return out
	}

	prev := seedSum / float64(period)
	out[period-1] = prev
	for i := period; i < len(data); i++ {
		prev = (data[i]-prev)*k + prev
		out[i] = prev
	}
	return out
}

// WMA returns the Weighted Moving Average.
func WMA(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	weightSum := float64(period*(period+1)) / 2.0

	for i := 0; i < len(data); i++ {
		if i < period-1 {
			out[i] = math.NaN()
			continue
		}
		sum := 0.0
		for j := 0; j < period; j++ {
			sum += data[i-j] * float64(period-j)
		}
		out[i] = sum / weightSum
	}
	return out
}

// RSI returns the Relative Strength Index (Wilder's smoothing).
func RSI(data []float64, period int) []float64 {
	if period <= 0 || len(data) < period+1 {
		return makeNaNFilled(len(data))
	}
	out := make([]float64, len(data))
	for i := 0; i < period; i++ {
		out[i] = math.NaN()
	}

	// Initial average gain/loss using SMA
	avgGain := 0.0
	avgLoss := 0.0
	for i := 1; i <= period; i++ {
		delta := data[i] - data[i-1]
		if delta > 0 {
			avgGain += delta
		} else {
			avgLoss += -delta
		}
	}
	avgGain /= float64(period)
	avgLoss /= float64(period)

	if avgLoss == 0 {
		out[period] = 100.0
	} else {
		rs := avgGain / avgLoss
		out[period] = 100.0 - (100.0 / (1.0 + rs))
	}

	// Wilder's smoothing for remaining values
	for i := period + 1; i < len(data); i++ {
		delta := data[i] - data[i-1]
		gain := 0.0
		loss := 0.0
		if delta > 0 {
			gain = delta
		} else {
			loss = -delta
		}
		avgGain = (avgGain*float64(period-1) + gain) / float64(period)
		avgLoss = (avgLoss*float64(period-1) + loss) / float64(period)

		if avgLoss == 0 {
			out[i] = 100.0
		} else {
			rs := avgGain / avgLoss
			out[i] = 100.0 - (100.0 / (1.0 + rs))
		}
	}
	return out
}

// StdDev returns the population standard deviation over `period` values.
func StdDev(data []float64, period int) []float64 {
	if period <= 1 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	for i := 0; i < len(data); i++ {
		if i < period-1 {
			out[i] = math.NaN()
			continue
		}
		// Calculate mean of window
		sum := 0.0
		for j := i - period + 1; j <= i; j++ {
			sum += data[j]
		}
		mean := sum / float64(period)

		// Calculate variance
		varianceSum := 0.0
		for j := i - period + 1; j <= i; j++ {
			diff := data[j] - mean
			varianceSum += diff * diff
		}
		out[i] = math.Sqrt(varianceSum / float64(period))
	}
	return out
}

// Bollinger returns Bollinger Bands (middle, upper, lower).
// Uses SMA for the middle band and `multiplier` standard deviations for the bands.
func Bollinger(data []float64, period int, multiplier float64) (middle, upper, lower []float64) {
	if period <= 1 || len(data) == 0 {
		return nil, nil, nil
	}
	middle = SMA(data, period)
	stddev := StdDev(data, period)
	upper = make([]float64, len(data))
	lower = make([]float64, len(data))

	for i := 0; i < len(data); i++ {
		if math.IsNaN(middle[i]) || math.IsNaN(stddev[i]) {
			upper[i] = math.NaN()
			lower[i] = math.NaN()
			continue
		}
		band := multiplier * stddev[i]
		upper[i] = middle[i] + band
		lower[i] = middle[i] - band
	}
	return
}

// ATR returns the Average True Range using Wilder's smoothing.
func ATR(high, low, close []float64, period int) []float64 {
	if period <= 0 || len(high) == 0 {
		return nil
	}
	n := len(high)
	out := make([]float64, n)

	// True range
	tr := make([]float64, n)
	tr[0] = high[0] - low[0]
	for i := 1; i < n; i++ {
		h := high[i]
		l := low[i]
		prevC := close[i-1]
		tr[i] = math.Max(h-l, math.Max(math.Abs(h-prevC), math.Abs(l-prevC)))
	}

	// Seed with SMA of first `period` TR values
	if n < period {
		return makeNaNFilled(n)
	}
	for i := 0; i < period-1; i++ {
		out[i] = math.NaN()
	}

	sum := 0.0
	for i := 0; i < period; i++ {
		sum += tr[i]
	}
	out[period-1] = sum / float64(period)

	// Wilder's smoothing
	for i := period; i < n; i++ {
		out[i] = (out[i-1]*float64(period-1) + tr[i]) / float64(period)
	}
	return out
}

// Stochastic returns the Stochastic Oscillator %K and %D lines.
func Stochastic(high, low, close []float64, kPeriod, kSlow, dPeriod int) (k, d []float64) {
	n := len(high)
	if kPeriod <= 0 || n < kPeriod {
		nf := makeNaNFilled(max(n, 0))
		return nf, nf
	}
	k = make([]float64, n)
	d = make([]float64, n)

	// Fast %K
	fastK := make([]float64, n)
	for i := 0; i < n; i++ {
		if i < kPeriod-1 {
			fastK[i] = math.NaN()
			continue
		}
		highestH := high[i]
		lowestL := low[i]
		for j := i - kPeriod + 1; j <= i; j++ {
			if high[j] > highestH {
				highestH = high[j]
			}
			if low[j] < lowestL {
				lowestL = low[j]
			}
		}
		denom := highestH - lowestL
		if denom == 0 {
			fastK[i] = 100.0
		} else {
			fastK[i] = ((close[i] - lowestL) / denom) * 100.0
		}
	}

	// Slow %K = SMA of fast %K
	k = SMA(fastK, kSlow)

	// %D = SMA of %K
	d = SMA(k, dPeriod)
	return
}

// VWAP returns the Volume-Weighted Average Price (cumulative from bar 0).
func VWAP(close, volume []float64) []float64 {
	n := len(close)
	out := make([]float64, n)
	cumPV := 0.0
	cumV := 0.0
	for i := 0; i < n; i++ {
		if volume[i] <= 0 {
			out[i] = math.NaN()
			continue
		}
		cumPV += close[i] * volume[i]
		cumV += volume[i]
		if cumV == 0 {
			out[i] = math.NaN()
		} else {
			out[i] = cumPV / cumV
		}
	}
	return out
}

// OBV returns On-Balance Volume.
func OBV(close, volume []float64) []float64 {
	n := len(close)
	if n == 0 {
		return nil
	}
	out := make([]float64, n)
	out[0] = float64(volume[0])
	for i := 1; i < n; i++ {
		if close[i] > close[i-1] {
			out[i] = out[i-1] + volume[i]
		} else if close[i] < close[i-1] {
			out[i] = out[i-1] - volume[i]
		} else {
			out[i] = out[i-1]
		}
	}
	return out
}

// MACD returns the MACD line, signal line, and histogram.
func MACD(data []float64, fast, slow, signal int) (macdLine, signalLine, histogram []float64) {
	n := len(data)
	if n < slow {
		nf := makeNaNFilled(max(n, 0))
		return nf, nf, nf
	}
	fastEMA := EMA(data, fast)
	slowEMA := EMA(data, slow)

	macdLine = make([]float64, n)
	for i := 0; i < n; i++ {
		if math.IsNaN(fastEMA[i]) || math.IsNaN(slowEMA[i]) {
			macdLine[i] = math.NaN()
		} else {
			macdLine[i] = fastEMA[i] - slowEMA[i]
		}
	}

	signalLine = EMA(macdLine, signal)
	histogram = make([]float64, n)
	for i := 0; i < n; i++ {
		if math.IsNaN(macdLine[i]) || math.IsNaN(signalLine[i]) {
			histogram[i] = math.NaN()
		} else {
			histogram[i] = macdLine[i] - signalLine[i]
		}
	}
	return
}

// Highest returns the highest value over the last `period` bars.
func Highest(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	for i := 0; i < len(data); i++ {
		if i < period-1 {
			out[i] = math.NaN()
			continue
		}
		maxV := data[i]
		for j := i - period + 1; j <= i; j++ {
			if data[j] > maxV {
				maxV = data[j]
			}
		}
		out[i] = maxV
	}
	return out
}

// Lowest returns the lowest value over the last `period` bars.
func Lowest(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	for i := 0; i < len(data); i++ {
		if i < period-1 {
			out[i] = math.NaN()
			continue
		}
		minV := data[i]
		for j := i - period + 1; j <= i; j++ {
			if data[j] < minV {
				minV = data[j]
			}
		}
		out[i] = minV
	}
	return out
}

// ROC returns the Rate of Change (percentage) over `period` bars.
func ROC(data []float64, period int) []float64 {
	if period <= 0 || len(data) == 0 {
		return nil
	}
	out := make([]float64, len(data))
	for i := 0; i < len(data); i++ {
		if i < period {
			out[i] = math.NaN()
			continue
		}
		prev := data[i-period]
		if prev == 0 {
			out[i] = math.NaN()
		} else {
			out[i] = ((data[i] - prev) / prev) * 100.0
		}
	}
	return out
}

// Crossover returns true when `fast` crosses above `slow` at the current bar.
// Both inputs are indicator arrays (may contain NaN for warm-up).
func Crossover(fast, slow []float64, i int) bool {
	if i < 1 || i >= len(fast) || i >= len(slow) {
		return false
	}
	return fast[i] > slow[i] &&
		!math.IsNaN(fast[i-1]) && !math.IsNaN(slow[i-1]) &&
		fast[i-1] <= slow[i-1]
}

// Crossunder returns true when `fast` crosses below `slow` at the current bar.
func Crossunder(fast, slow []float64, i int) bool {
	if i < 1 || i >= len(fast) || i >= len(slow) {
		return false
	}
	return fast[i] < slow[i] &&
		!math.IsNaN(fast[i-1]) && !math.IsNaN(slow[i-1]) &&
		fast[i-1] >= slow[i-1]
}

func makeNaNFilled(n int) []float64 {
	out := make([]float64, n)
	for i := range out {
		out[i] = math.NaN()
	}
	return out
}

File: haven-desktop/frontend/src/sdk/indicators.js
// Indicator library — plain functions, array-in / array-out, `null` during the
// warm-up period so strategies can guard with `x[i] == null`. No dependencies.
// Extending: export one more function here and wire it into the ctx table in
// runtime.js (one line each).

export function sma(src, len) {
  const out = new Array(src.length).fill(null);
  let sum = 0;
  for (let i = 0; i < src.length; i++) {
    sum += src[i];
    if (i >= len) sum -= src[i - len];
    if (i >= len - 1) out[i] = sum / len;
  }
  return out;
}

export function ema(src, len) {
  const out = new Array(src.length).fill(null);
  const k = 2 / (len + 1);
  let prev = null;
  let seed = 0;
  for (let i = 0; i < src.length; i++) {
    if (i < len - 1) { seed += src[i]; continue; }
    if (i === len - 1) { prev = (seed + src[i]) / len; }       // seed with SMA
    else { prev = src[i] * k + prev * (1 - k); }
    out[i] = prev;
  }
  return out;
}

export function wma(src, len) {
  const out = new Array(src.length).fill(null);
  const denom = (len * (len + 1)) / 2;
  for (let i = len - 1; i < src.length; i++) {
    let sum = 0;
    for (let j = 0; j < len; j++) sum += src[i - j] * (len - j);
    out[i] = sum / denom;
  }
  return out;
}

// Wilder's RSI.
export function rsi(src, len) {
  const out = new Array(src.length).fill(null);
  let avgGain = 0, avgLoss = 0;
  for (let i = 1; i < src.length; i++) {
    const change = src[i] - src[i - 1];
    const gain = Math.max(change, 0);
    const loss = Math.max(-change, 0);
    if (i <= len) {
      avgGain += gain / len;
      avgLoss += loss / len;
      if (i < len) continue;
    } else {
      avgGain = (avgGain * (len - 1) + gain) / len;
      avgLoss = (avgLoss * (len - 1) + loss) / len;
    }
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

export function macd(src, fastLen = 12, slowLen = 26, signalLen = 9) {
  const fast = ema(src, fastLen);
  const slow = ema(src, slowLen);
  const macdLine = src.map((_, i) =>
    fast[i] != null && slow[i] != null ? fast[i] - slow[i] : null);
  // Signal = EMA of the macd line, starting where it becomes defined.
  const start = macdLine.findIndex(v => v != null);
  const signal = new Array(src.length).fill(null);
  if (start !== -1) {
    const seg = ema(macdLine.slice(start), signalLen);
    for (let i = 0; i < seg.length; i++) signal[start + i] = seg[i];
  }
  const hist = macdLine.map((v, i) =>
    v != null && signal[i] != null ? v - signal[i] : null);
  return { macd: macdLine, signal, hist };
}

export function stddev(src, len) {
  const out = new Array(src.length).fill(null);
  const mean = sma(src, len);
  for (let i = len - 1; i < src.length; i++) {
    let sumSq = 0;
    for (let j = i - len + 1; j <= i; j++) sumSq += (src[j] - mean[i]) ** 2;
    out[i] = Math.sqrt(sumSq / len);
  }
  return out;
}

export function bollinger(src, len = 20, mult = 2) {
  const middle = sma(src, len);
  const sd = stddev(src, len);
  const upper = middle.map((m, i) => (m != null ? m + mult * sd[i] : null));
  const lower = middle.map((m, i) => (m != null ? m - mult * sd[i] : null));
  return { upper, middle, lower };
}

// Wilder's ATR.
export function atr(high, low, close, len = 14) {
  const out = new Array(close.length).fill(null);
  let prev = null;
  let seed = 0;
  for (let i = 0; i < close.length; i++) {
    const tr = i === 0
      ? high[i] - low[i]
      : Math.max(high[i] - low[i], Math.abs(high[i] - close[i - 1]), Math.abs(low[i] - close[i - 1]));
    if (i < len - 1) { seed += tr; continue; }
    if (i === len - 1) prev = (seed + tr) / len;
    else prev = (prev * (len - 1) + tr) / len;
    out[i] = prev;
  }
  return out;
}

export function stochastic(high, low, close, kLen = 14, dLen = 3) {
  const k = new Array(close.length).fill(null);
  for (let i = kLen - 1; i < close.length; i++) {
    let hh = -Infinity, ll = Infinity;
    for (let j = i - kLen + 1; j <= i; j++) {
      if (high[j] > hh) hh = high[j];
      if (low[j] < ll) ll = low[j];
    }
    k[i] = hh === ll ? 50 : ((close[i] - ll) / (hh - ll)) * 100;
  }
  // %D = SMA of %K over its defined segment.
  const d = new Array(close.length).fill(null);
  const start = k.findIndex(v => v != null);
  if (start !== -1) {
    const seg = sma(k.slice(start), dLen);
    for (let i = 0; i < seg.length; i++) d[start + i] = seg[i];
  }
  return { k, d };
}

// Cumulative (session-less) VWAP over the whole series.
export function vwap(high, low, close, volume) {
  const out = new Array(close.length).fill(null);
  let pv = 0, v = 0;
  for (let i = 0; i < close.length; i++) {
    const typical = (high[i] + low[i] + close[i]) / 3;
    pv += typical * volume[i];
    v += volume[i];
    out[i] = v > 0 ? pv / v : null;
  }
  return out;
}

export function obv(close, volume) {
  const out = new Array(close.length).fill(null);
  let acc = 0;
  for (let i = 0; i < close.length; i++) {
    if (i > 0) {
      if (close[i] > close[i - 1]) acc += volume[i];
      else if (close[i] < close[i - 1]) acc -= volume[i];
    }
    out[i] = acc;
  }
  return out;
}

export function highest(src, len) {
  const out = new Array(src.length).fill(null);
  for (let i = len - 1; i < src.length; i++) {
    let hh = -Infinity;
    for (let j = i - len + 1; j <= i; j++) if (src[j] > hh) hh = src[j];
    out[i] = hh;
  }
  return out;
}

export function lowest(src, len) {
  const out = new Array(src.length).fill(null);
  for (let i = len - 1; i < src.length; i++) {
    let ll = Infinity;
    for (let j = i - len + 1; j <= i; j++) if (src[j] < ll) ll = src[j];
    out[i] = ll;
  }
  return out;
}

// Rate of change, percent.
export function roc(src, len) {
  const out = new Array(src.length).fill(null);
  for (let i = len; i < src.length; i++) {
    if (src[i - len] !== 0) out[i] = ((src[i] - src[i - len]) / src[i - len]) * 100;
  }
  return out;
}

// a crossed above/below b at index i. Either side may be an array or a constant.
const at = (s, i) => (Array.isArray(s) ? s[i] : s);

export function crossover(a, b, i) {
  if (i < 1) return false;
  const a0 = at(a, i - 1), a1 = at(a, i), b0 = at(b, i - 1), b1 = at(b, i);
  if (a0 == null || a1 == null || b0 == null || b1 == null) return false;
  return a0 <= b0 && a1 > b1;
}

export function crossunder(a, b, i) {
  if (i < 1) return false;
  const a0 = at(a, i - 1), a1 = at(a, i), b0 = at(b, i - 1), b1 = at(b, i);
  if (a0 == null || a1 == null || b0 == null || b1 == null) return false;
  return a0 >= b0 && a1 < b1;
}
